### ISSUE TITLE
error on Ubuntu 20.04

### ISSUE THREAD

**Issue by FATeknollogee** [Ylianst/MeshCentral#1323]
"The system does not appear to have an XServer instance running when no users are logged in. KVM child process has unexpectedly exited"

New install on Ubuntu.
I get the above when I try to connect to the desktop.

What's the fix?

**Comment by krayon007** [comment]
When you try to connect the desktop, are any users logged in?

**Comment by FATeknollogee** [comment]
No users are logged in.

**Comment by krayon007** [comment]
If no users are logged in, that is why it cannot connect. Newer version of X have a "security" feature where the server instance is not spawned until a user logs in. That means when no users are logged in, there is no X server instance for the agent to connect to. (Mesh Central relies on X to do screen capture, so that is doesn't require any custom software/video drivers/etc.) Which unfortunately means, that on these platforms, it is not possible to do screen capture until a user logs in.

**Comment by FATeknollogee** [comment]
Thx for the heads up!

**Comment by FATeknollogee** [comment]
@krayon007 I assume there is no work around to this "no user logged in" issue?

**Comment by krayon007** [comment]
I tried to find one, but have been unsuccessful. The closest I got, was to spawn a login shell, and ask for a username/password, but that spawned a non graphical login. I tried to run X manually that way, but the desktop that came up was very different than the actual desktop.

**Comment by Ylianst** [comment]
Is there any way to display a agent console message indicating that the login screen can't be captured? I imagine that would help lots with users.

**Comment by guerby** [comment]
@krayon007 gdm3 actually runs an X server with -xauth for the login screen with user "gdm", on my ubuntu 20.04 test system : 

```
root       56488  0.0  0.0 175412  9164 ?        Sl   mai07   0:00  \_ gdm-session-worker [pam/gdm-launch-environment]
gdm        56497  0.0  0.0 172804  6844 tty1     Ssl+ mai07   0:00  |   \_ /usr/lib/gdm3/gdm-x-session gnome-session --autostart /usr/share/gdm/greeter/autostart
gdm        56499  0.0  0.1 805492 61912 tty1     Sl+  mai07   0:13  |       \_ /usr/lib/xorg/Xorg vt1 -displayfd 3 -auth /run/user/125/gdm/Xauthority -background none -noreset -keeptty -verbose 3
gdm        56528  0.0  0.0 273548 15916 tty1     Sl+  mai07   0:00  |       \_ /usr/libexec/gnome-session-binary --systemd --autostart /usr/share/gdm/greeter/autostart
```

Process 56499 is a regular X server, and user "gdm" has a uid:

```
grep gdm /etc/passwd
gdm:x:125:130:Gnome Display Manager:/var/lib/gdm3:/bin/false
```

I did as root while the login screen was running:

```
DISPLAY=:0 XAUTHORITY=/run/user/125/gdm/Xauthority xwd -root > z.xwd
apt-get install imagemagick
convert z.xwd z.png
```

and I got (visual keyboard activated) : 

![z](https://user-images.githubusercontent.com/6659329/81569159-32d31980-939f-11ea-8d6b-5d4ca354f58d.png)

So capture of the login screen is definitely possible (with WaylandEnable=false in /etc/gdm3/custom.conf )

**Comment by guerby** [comment]
I don't know why meshagent doesn't pick up the gdm login screen.

Also most Linux systems now allow multiple users with their own session (one active, the other are lock screen until explicit user switch).

**Comment by guerby** [comment]
@Ylianst more or les the same bug report may be to be joined :  
https://github.com/Ylianst/MeshCentral/issues/1287

**Comment by krayon007** [comment]
ok, I think I see what's going on with Ubuntu 20. When no users are logged in, the agent looks for an xserver running as root. On Ubuntu 20, when nobody is logged in, the xserver is running as gdm.  I'll have to modify the checks, to look for the uid of the display manager user, if there is one. That means it should be possible to place a toast, connection bar, and user-consent on the login screen too, on Ubuntu 20

**Comment by guerby** [comment]
On Debian 10 (buster) the X process run as user "Debian-gdm"

```
Debian-gdm:x:116:122:Gnome Display Manager:/var/lib/gdm3:/bin/false
```

**Comment by krayon007** [comment]
I have a bunch of different test VM's set up with all the major distros, so I'm going to try to enable it in a broad as fashion as possible....  I'm going to modify it so that when the agent asks for the console id, It will check to see if the xserver is running if no users are logged in, and if so, return the uid of the display manager. That way, desktop, toast, and user-consent should all work from the login screen without any code changed.

**Comment by krayon007** [comment]
Ok, I got it working... Some distros work a little differently than others... openSUSE for example (which apparently always worked), when you login from the login screen, it seamlessly transitions to the logged in user, I think because the DISPLAY port does not change..

Fedora and Ubuntu 20,  after you login, the screen turns black, and you have to disconnect and reconnect, and it will connect correctly... I think this is becuase the display port changes from :1024 to :0 when you login...

I'll work on seeing if I can detect the change and switch over automatically... But here's a screen shot of it working:

![login](https://user-images.githubusercontent.com/29240564/81613118-bcc8b580-9392-11ea-976a-71e416993a2b.png)

**Comment by krayon007** [comment]
On a side note, on Ubuntu 20, the default install uses XWayland for the login process, which does not work with our KVM. (It connects, but does not receive any data from XWayland)... To get around this, on Ubuntu 20, you'll need to edit /etc/gdm3/custom.conf and disable XWayland, and enable Xorg...

But it looks like you already did that, based on the screencapture you posted earlier where you showed the PIDs.

**Comment by krayon007** [comment]
Also, it looks like user-consent, privacy bar, and toasts do not work on the login screen. I think it might be because the GDM session for the login shell does not allow the xclient to draw any windows.

**Comment by guerby** [comment]
> On a side note, on Ubuntu 20, the default install uses XWayland for the login process, which does not work with our KVM. (It connects, but does not receive any data from XWayland)... To get around this, on Ubuntu 20, you'll need to edit /etc/gdm3/custom.conf and disable XWayland, and enable Xorg...
> 
> But it looks like you already did that, based on the screencapture you posted earlier where you showed the PIDs.

Yes I deactivated wayland in gdm3 for my meshcentral tests.

**Comment by guerby** [comment]
@krayon007 Hi, I tried again today with MeshCentral 0.5.45,  versions on console says "meshAgent": "2020-05-20T19:22:09.000Z", MeshCore CRC-100443630 and could not get the login banner on my test systems running respectively debian 10, ubuntu 20.04, Fedora 32 (all fully up to date, and with WaylandEnable=false, freshly rebooted).

Symptom is when I click desktop / connect it goes back to disconnected immediately and desktop stays black.

May be your patches are not included in 0.5.45 yet?

**Comment by krayon007** [comment]
In addition to WaylandEnable=false
did you also add the following underneath it?
DefaultSession=gnome-xorg.desktop

**Comment by krayon007** [comment]
Also, do you have user-consent enabled? User-consent cannot work on the login screen, so if you attempt to connect to the desktop that is sitting on the login screen, with user-consent enabled, it will auto-deny the connection. (You should see a message saying it was denied tho)

**Comment by krayon007** [comment]
I just tested it on Ubuntu 20, with the current version of Mesh Central, and it works correctly, so I'm hoping it's one of the issues mentioned above.

**Comment by FATeknollogee** [comment]
Ubuntu 20 doesn't work for me. All I get is Connected to a black screen.

**Comment by FATeknollogee** [comment]
@krayon007 Is this line necessary (DefaultSession=gnome-xorg.desktop) or is that more of a "just in case"?

**Comment by guerby** [comment]
@krayon007 I now understand the issue: I use physical Intel NUCs for my tests and they didn't have a monitor plugged on their HDMI port. 

If I plug an HDMI monitor meshcentral works, as soon as I unplug it ceases to work (meshcentral desktop image freezes, and disconnect/connect gets black screen then disconnect again). 

The same thing happens on my three test machines which are different models of NUC.

Note that if just turning off the monitor (with the monitor on/off control button, monitor still plugged to AC) then meshcentral works.

I don't know how Xorg reacts to monitor / no monitor plugged in but it's probably impacting meshagent in some way.

**Comment by krayon007** [comment]
Oh, ok, I remember seeing that issue with our Raspberry Pis. The issue is because when no physical monitor is plugged in, the X server returns to me an empty list, when I ask for all the displays, so I have no display ID to use, which I need to specify when I scrape the desktop. When I get some more time, I'll look into seeing how to enumerate virtual displays/desktops. I was able to enumerate virtual workspaces, but I still need a Display ID to enumerate that.

**Comment by krayon007** [comment]
I'll add a check for that, so in the future, I can display a message asking if any monitors are connected or something similar...

**Comment by krayon007** [comment]
> @krayon007 Is this line necessary (DefaultSession=gnome-xorg.desktop) or is that more of a "just in case"?

I don't know actually. I got it from some documentation years ago... I just always put them together. I think it was because with one of the old setups I had (I honestly don't even remember), kvm didn't work unless I specified this. But this was way before I updated the KVM to be smarter with how to find the DISPLAY and XAUTHORITY strings, so it's entirely possible that it's not necessary.

**Comment by FATeknollogee** [comment]
Ubuntu 20.04 desktop (not a VM) with no monitor hooked up, I still get a "connected" black screen.
Fedora 32 xfce desktop (not a vm) with no monitor connects ok.

**Comment by krayon007** [comment]
This is a known issue when no monitor is connected. I didnt' know there were systems that actually worked in this scenario, but now I'll take a closer look. In the past, I know that when no physical monitors were connected, X would give me an empty list when I enumerated all the displays. Without a valid dispay ID, X will not let me scrape the desktop. I meant to update the error messages, so relay an error if X gives me an empty list. When I free up more time, I'll look into trying to figure out how to get a display ID when there are no monitors connected, on distros where this happens.

**Comment by guerby** [comment]
I found out how to have a working meshcentral without a physical monitor plugged in:

Ubuntu & debian: `apt-get install xserver-xorg-video-dummy`
Fedora: `dnf install xorg-x11-drv-dummy`

Add file  `/usr/share/X11/xorg.conf.d/10-headless.conf`  with content:

```
Section "Monitor"
    Identifier "dummy_monitor"
    HorizSync 28.0-80.0
    VertRefresh 48.0-75.0
    Modeline "1920x1080" 172.80 1920 2040 2248 2576 1080 1081 1084 1118
EndSection

Section "Device"
    Identifier "dummy_card"
    VideoRam 256000
    Driver "dummy"
EndSection

Section "Screen"
    Identifier "dummy_screen"
    Device "dummy_card"
    Monitor "dummy_monitor"
    SubSection "Display"
        Modes "1920x1080"
    EndSubSection
EndSection
```

Hat tip to https://wiki.archlinux.org/index.php/Xorg#Manual_configuration

It works on ubuntu 20.04 and Fedora 32 but on debian 10 meshagent says on stderr: "No protocol specified" which probably means it didn't pickup Xauthority correctly, if I use xwd -root I get the gdm3 login page. The Xorg process commandline looks normal:

```
Debian-+   902  0.1  0.3 761432 59152 tty1     Sl+  13:47   0:00          \_ /usr/lib/xorg/Xorg vt1 -displayfd 3 -auth /run/user/116/gdm/Xauthority -background none -noreset -keeptty -verbose 3
```
xrandr is also ok:

```
# XAUTHORITY=/run/user/116/gdm/Xauthority DISPLAY=:0 xrandr -q
Screen 0: minimum 320 x 240, current 1920 x 1080, maximum 1920 x 1080
default connected primary 1920x1080+0+0 0mm x 0mm
```

I don't know why meshagent isn't picking it up on debian 10 while it does on ubuntu 20.04 and fedora 32.

**Comment by FATeknollogee** [comment]
Yes, sir, thank you.
That works on ubuntu 20.04...!!

Do we still need this part: "edit /etc/gdm3/custom.conf and disable XWayland"?
I'll comment that out & see if it still works.

**Comment by FATeknollogee** [comment]
On ubuntu 20.04, after this (see below), it still works.
```
#WaylandEnable=false
#DefaultSession=gnome-xorg.desktop

**Comment by krayon007** [comment]
> It works on ubuntu 20.04 and Fedora 32 but on debian 10 meshagent says on stderr: "No protocol specified" which probably means it didn't pickup Xauthority correctly, if I use xwd -root I get the gdm3 login page. The Xorg process commandline looks normal:
> 
> ```
> Debian-+   902  0.1  0.3 761432 59152 tty1     Sl+  13:47   0:00          \_ /usr/lib/xorg/Xorg vt1 -displayfd 3 -auth /run/user/116/gdm/Xauthority -background none -noreset -keeptty -verbose 3
> ```
> 
> 
> I don't know why meshagent isn't picking it up on debian 10 while it does on ubuntu 20.04 and fedora 32.

Assuming nobody is logged in, can you connect a terminal, and then do the following command, and let me know what it says?
./meshagent -exec "console.log(require('monitor-info').getXInfo(116));process.exit();"

**Comment by krayon007** [comment]
And can you also run this command too:
./meshagent -exec "console.log(require('user-sessions').gdmUid);process.exit();"

**Comment by guerby** [comment]
@krayon007 here are the results:

```
# /usr/local/mesh/meshagent -exec "console.log(require('monitor-info').getXInfo(116));process.exit();"
{ tty: tty1, xauthority: /run/user/116/gdm/Xauthority, exportEnv: function exportEnv() { [ecmascript code] }, display: :0, xdm: xorg }
# /usr/local/mesh/meshagent -exec "console.log(require('user-sessions').gdmUid);process.exit();"
116
```

**Comment by krayon007** [comment]
So it looks like the Mesh Agent is able to correctly get the xauthority on your Debian 10 system... Something else must be going on...

Can you edit your: /lib/systemd/system/meshagent.service

edit the following line, by adding the --slaveKvmLog="1", like the following:
ExecStart=/usr/local/mesh_services/meshagent/meshagent --slaveKvmLog="1"

Then restart the service. 
The meshagent will write some logs regarding it's X connection to: /tmp/slave

In the meantime, I'll see if I can install Debian 10 onto something that isn't a laptop. I tested with a NUC running Peppermint 9, but it seems to connect just fine with no displays attached.

**Comment by guerby** [comment]
@krayon007 here is the result:
```
# cat /tmp/slave 
SLAVE/KVM CoreDumps DISABLED
Checking $DISPLAY
XAUTHORITY is (null)
DisplayString is :0
```

**Comment by krayon007** [comment]
Ok, that's wierd. That XAUTHORITY should have been set, because it was obtained by the parent process, using the command you ran earlier (getXInfo).

On a side not, I found a scenario where the KVM will connect to a black screen on the login screen. If you connect a terminal login shell, and login, the KVM will not connect correctly, becuase it will see that user 1000 (or whatever user logged in with the login shell), but it will fail to connect, because that user session does not have a graphical session.  I'll have to work on that situation...

On your system, can you connect a root terminal session while you are at the login screen, and type the following command:

./meshagent -exec "try{console.log(require('user-sessions').consoleUid());}catch(e){console.log(e);}process.exit();"

(That should all be on one line)

**Comment by guerby** [comment]
@krayon007 here is what I get:

```
# /usr/local/mesh/meshagent -exec "try{console.log(require('user-sessions').consoleUid());}catch(e){console.log(e);}process.exit();"
0
```

To be sure there was nothing broken on the machine I disabled 10-headless.conf and reboot with HDMI pluged in and everything worked 

```
# cat /tmp/slave 
SLAVE/KVM CoreDumps DISABLED
Checking $DISPLAY
ENV[DISPLAY] = 0
XAUTHORITY is /run/user/116/gdm/Xauthority
DisplayString is :0
```

Unplugged HDMI and put 10-headless.conf back and I get same not working result for meshcentral, got a gdm3 login screen capture with ` XAUTHORITY=/run/user/116/gdm/Xauthority DISPLAY=:0 xwd -root > zz3.xwd`

**Comment by krayon007** [comment]
Ok, one more test... While at the login screen, can you connect a root terminal, and run the who command?

I think the problem, is that when you called consoleUid(), it returned 0, instead of the gdmUid of 116. The only way it would do that, is if who returned that somebody was logged in.

**Comment by guerby** [comment]
@krayon007 as I'm logged as root with ssh from my main machine the who command outputs:

```
# who
root     pts/0        2020-06-01 01:16 (192.168.73.102)
```

**Comment by krayon007** [comment]
So basically, it's a similar issue as when I had a login shell via terminal... I'll have to modify the logic, so that it checks if the logged in user has a graphical session first, before it tries running with it, defaulting to the gdmuid if it doesn't have a graphical session...

I'm not sure why you didn't have this issue on your other machines... Did you not have an ssh session connection to those machines?

**Comment by krayon007** [comment]
Although, I'm not sure why it didn't also return 0, when you had an actual HDMI monitor plugged in....

**Comment by guerby** [comment]
@krayon007 I just checked and on the ubuntu and fedora machines who outputs the exact same line as I'm logged in as root with ssh.

```
root@nuc2:~# who
root     pts/0        2020-06-01 01:50 (192.168.73.102)

[root@nuc3 ~]# who
root     pts/0        2020-06-01 01:53 (192.168.73.102)
```

**Comment by guerby** [comment]
@krayon007 good catch ! I disconnected from ssh and meshcentral on debian started to work. I connected with ssh to the ubuntu machines and it ceased to work, so the issue is likely to be ssh session breaking meshagent logic.

**Comment by krayon007** [comment]
Ok, I modified user-sessions, so when you query for the logged in user, it will ignore users logged in  on a pseudo terminal device, which will fix this issue, as the KVM will no longer try to KVM in this scenario where you are connected via SSH/Login Shell, etc. This will be in the next agent update.

**Comment by guerby** [comment]
@krayon007 I updated my test server to 0.5.72 and it seems to work even when ssh session are present (debian 10 / ubuntu 20.04 / fedora32). I assume your fixes are released now?

**Comment by krayon007** [comment]
> @krayon007 I updated my test server to 0.5.72 and it seems to work even when ssh session are present (debian 10 / ubuntu 20.04 / fedora32). I assume your fixes are released now?

Yes. This particular fix was in v0.5.51

**Comment by NiceGuyIT** [comment]
@si458 Please close. This was fixed in 2020 (v0.5.51).
